README.md

```diff
@@ -1,38 +1,104 @@
+constants by arithmetic are also considered hard-coded. For example `1 + x` is a
+hard-coded constant, even if `x` is not. This rule is conservative: it might overestimate
+which values in the program are hard-coded. It is necessary, though, because an
+expression like `x * 0` really is hard-coded, even if `x` is not.
+
+Note that `boolean`s are *not* considered constants, because they only have two values
+and are not interesting for the security properties this checker is meant to track.
+`null` is also *not* considered a constant; for null-tracking, see the [Nullness
+Checker](https://checkerframework.org/manual/#nullness-checker) of the Checker Framework.
+
+### Annotating your program
+
+The checker assigns each expression in the program one of
+these two types:
+
+```java
+@MaybeDerivedFromConstant
+            |
+       @NonConstant
+```
+
+The default is `@NonConstant` for all expressions except manifest literals.
+Because method parameters default to `@NonConstant`, the user must write a `@MaybeDerivedFromConstant` annotation on the definition
+of any method that may be called with a constant argument. For example, if your code includes
+the correct call:
+```java
+foo(5);
+```
+then the defintion of `foo` will need to be annotated like so:
+```java
+void foo(@MaybeDerivedFromConstant int x) { }
+```
+
+### You must annotate libraries
+
+For unchecked code (that is, code for which
+only the bytecode is available, such as code from a `.jar` file), the following
+optimistic defaulting rules are applied:
+* the return type of any function defined in a library is `@NonConstant` if the function takes no parameters, or all of its parameters are non-constant
+* the types of a library function's formal parameters are always `@MaybeDerivedFromConstant`
+
+unless a stub file is supplied that overwrites this default
+
+As a consequence of these rules, users MUST always write stub files for libraries
+they intend to protect -- this checker is **useless** without such a stub.
+
+For example, [one of the No Literal Checker's built-in stub
+files](https://github.com/kelloggm/no-literal-checker/blob/prototype/no-literal-checker/stubs/crypto.astub)
+enforces that calls to the `javax.crypto.spec.SecretKeySpec` constructor only
+ever provide non-constant keys:
+
+```java
+package javax.crypto.spec;
+
+import org.checkerframework.checker.noliteral.qual.NonConstant;
+
+class SecretKeySpec {
+    SecretKeySpec(@NonConstant byte[] key, String algorithm);
+    SecretKeySpec(@NonConstant byte[] key, int offset, int len, String algorithm);
+}
+```
+
+For more about stub files, see the
+[Checker Framework manual](https://checkerframework.org/manual/#stub).
+
+### Local type inference for array component types
+
+Unlike most typecheckers built on the Checker Framework, this checker includes custom rules
+for inferring the component types of local arrays. That is, the checker can infer that
```

Comment:
Is this only the innermost component type, or any component type (which might itself be an array, as a component [in a multi-dimensional array)?

Reply:
Only the innermost component type. Array types are not considered possible constants by the checker - only the innermost component type matters. I'll update this documentation.

Reply:
Related, an empty array isn't really an exception: its "components" should be treated as the constant, not the array itself (because they're always the same - nothing!). The type of `T[] = new T[0]` should be `@MaybeDerivedFromConstant T @NonConstant []`, perhaps somewhat counter-intuitively.